File: app/src/services/rag.ts
import { invoke } from '@tauri-apps/api/core';

export interface RagBuildSummary {
  available: boolean;
  model: string;
  schemaVersion: string;
  error?: string;
  builtAt?: string | null;
  filesIndexed: number;
  chunksIndexed: number;
  filesScanned: number;
  filesUpdated: number;
  filesRemoved: number;
}

export interface RagSearchHit {
  path: string;
  size: number;
  outline: string;
  chunkType: string;
  title?: string | null;
  startLine: number;
  endLine: number;
  snippet: string;
  semanticScore: number;
  lexicalScore: number;
  combinedScore: number;
  supportingMatches: number;
}

export interface RagSearchResult {
  available: boolean;
  model: string;
  error?: string;
  builtAt?: string | null;
  filesIndexed: number;
  chunksIndexed: number;
  hits: RagSearchHit[];
}

function isTauriRuntime() {
  return typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window;
}

function formatInvokeError(error: unknown): string {
  if (error instanceof Error) return error.message;
  return String(error);
}

export async function rebuildWorkspaceRagIndex(workspacePath: string): Promise<RagBuildSummary | null> {
  if (!isTauriRuntime()) return null;
  try {
    return await invoke<RagBuildSummary>('rag_rebuild_index', { workspacePath });
  } catch (error) {
    const message = formatInvokeError(error);
    console.error('[RAG] rebuild failed:', message);
    return {
      available: false,
      model: 'AllMiniLML6V2',
      schemaVersion: '1',
      error: message,
      filesIndexed: 0,
      chunksIndexed: 0,
      filesScanned: 0,
      filesUpdated: 0,
      filesRemoved: 0,
    };
  }
}

export async function searchWorkspaceRag(
  workspacePath: string,
  query: string,
  options: {
    limit?: number;
    activeFile?: string;
    recentFiles?: string[];
  } = {},
): Promise<RagSearchResult | null> {
  if (!isTauriRuntime()) return null;
  try {
    return await invoke<RagSearchResult>('rag_search', {
      workspacePath,
      query,
      limit: options.limit ?? 10,
      activeFile: options.activeFile,
      recentFiles: options.recentFiles ?? [],
    });
  } catch (error) {
    const message = formatInvokeError(error);
    console.error('[RAG] search failed:', message);
    return {
      available: false,
      model: 'AllMiniLML6V2',
      error: message,
      filesIndexed: 0,
      chunksIndexed: 0,
      hits: [],
    };
  }
}

export async function releaseWorkspaceRagResources(): Promise<void> {
  if (!isTauriRuntime()) return;
  try {
    await invoke('rag_release_resources');
  } catch (error) {
    console.error('[RAG] release failed:', formatInvokeError(error));
  }
}

File: app/src-tauri/src/rag.rs
use serde::Serialize;

const RAG_MODEL_NAME: &str = "AllMiniLML6V2";
const RAG_SCHEMA_VERSION: &str = "2";

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RagBuildSummary {
    pub available: bool,
    pub model: &'static str,
    pub schema_version: &'static str,
    pub built_at: Option<String>,
    pub files_indexed: usize,
    pub chunks_indexed: usize,
    pub files_scanned: usize,
    pub files_updated: usize,
    pub files_removed: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RagSearchHit {
    pub path: String,
    pub size: u64,
    pub outline: String,
    pub chunk_type: String,
    pub title: Option<String>,
    pub start_line: usize,
    pub end_line: usize,
    pub snippet: String,
    pub semantic_score: f32,
    pub lexical_score: f32,
    pub combined_score: f32,
    pub supporting_matches: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RagSearchResult {
    pub available: bool,
    pub model: &'static str,
    pub built_at: Option<String>,
    pub files_indexed: usize,
    pub chunks_indexed: usize,
    pub hits: Vec<RagSearchHit>,
}

pub fn rebuild_index(_workspace_path: &str) -> Result<RagBuildSummary, String> {
    Ok(RagBuildSummary {
        available: false,
        model: RAG_MODEL_NAME,
        schema_version: RAG_SCHEMA_VERSION,
        built_at: None,
        files_indexed: 0,
        chunks_indexed: 0,
        files_scanned: 0,
        files_updated: 0,
        files_removed: 0,
    })
}

pub fn search_workspace(
    _workspace_path: &str,
    _query: &str,
    _limit: usize,
    _active_file: Option<&str>,
    _recent_files: &[String],
) -> Result<RagSearchResult, String> {
    Ok(RagSearchResult {
        available: false,
        model: RAG_MODEL_NAME,
        built_at: None,
        files_indexed: 0,
        chunks_indexed: 0,
        hits: Vec::new(),
    })
}

pub fn release_resources() -> Result<(), String> {
    Ok(())
}
